### PR TITLE
Fix FlashInfer A2A top-k ID dtype

### PR DESCRIPTION
**Pull request by nvpohanh**
[by Codex]

## Summary

- Materialize bypassed top-k output before FlashInfer MoE all-to-all dispatch.
- Convert selected expert IDs to `torch.int32`, as required by FlashInfer's `MoeAlltoAll.dispatch` API.
- Add a registered four-GPU B200 GLM-5.2 NVFP4 test covering both `DEP4+FlashInferA2A+StaticEP` and the default DEP4 backend.

## Motivation

SGLang's static expert-parallel dispatch remaps selected experts through an `int64` expert-location table. FlashInfer 0.6.12 requires `int32` expert IDs, so GLM-5.2 NVFP4 with DEP4, FlashInfer A2A, and `--ep-dispatch-algorithm=static` aborts during CUDA graph capture with:

```text
Check failed: tokenSelectedExperts.dtype() == dl_int32 (int64 vs. int32)
```

Normalizing the IDs at the FlashInfer dispatcher boundary keeps SGLang's internal routing representation independent of FlashInfer's ABI.

## FlashInfer ABI

SGLang pins FlashInfer 0.6.12. [`MoeAlltoAll.dispatch`](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.12/flashinfer/comm/trtllm_moe_alltoall.py#L596-L625) forwards `token_selected_experts` to the underlying API. Its [`moe_a2a_dispatch` docstring](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.12/flashinfer/comm/trtllm_moe_alltoall.py#L54-L72) specifies that `token_selected_experts` is an `int32` tensor.

## B200 verification

Ran the exact registered test in `test/registered/4-gpu-models/test_glm52_nvfp4.py` with four B200 GPUs and `nvidia/GLM-5.2-NVFP4`:

- Before this PR (`26b15a0825e32ff9fef4ff58b253f1c40a942981`): `DEP4+FlashInferA2A+StaticEP` deterministically failed during CUDA graph capture with the `int64` versus `int32` check above.
- With this PR (`478f904912`): the full combined test exited 0 in 1767 seconds with `OVERALL: ALL TESTS PASSED`.
  - `DEP4+FlashInferA2A+StaticEP`: performance PASS (804.1 output tok/s at batch size 16), GSM8K PASS (`0.958` versus `0.920` baseline).
  - `DEP4-default`: performance PASS (861.9 output tok/s at batch size 16), GSM8K PASS (`0.955` versus `0.920` baseline).

Additional checks:

- `python3 -m black --check test/registered/4-gpu-models/test_glm52_nvfp4.py`
- `python3 -m py_compile test/registered/4-gpu-models/test_glm52_nvfp4.py`
- `git diff --check 26b15a0825e32ff9fef4ff58b253f1c40a942981..HEAD`

---
### CI States

Latest PR Test (Base): :x: [Run #29388844304](https://github.com/sgl-project/sglang/actions/runs/29388844304)
Latest PR Test (Extra): :x: [Run #29388844184](https://github.com/sgl-project/sglang/actions/runs/29388844184)

### CODE REVIEW

**Comment by nvpohanh** [comment]
[by Codex] Addressed the review concern in ffaa7c6a4f: FlashInfer A2A now materializes `BypassedTopKOutput` before casting IDs, and the regression test covers both standard and bypassed top-k carriers. The focused test passed in `lmsysorg/sglang:nightly-dev-cu13-20260701-a01afdd5` (2 tests, 0 failures; OCI job 168393).

**Comment by nvpohanh** [comment]
[by Codex]
/tag-and-rerun-ci

**Comment by nvpohanh** [comment]
@leejnau @nvjullin @trevor-m Agents told me that this patch is needed to run GLM-5 + FlashInfer A2A. Do you think it makes sense?

**Comment by nvjullin** [comment]
I'm not that familiar with moe dispatch. However, I'm concerned that the conversion to int32 is unconditional. Did the dispatch ever work for other backend combinations? Wouldn't those break?

**Comment by nvpohanh** [comment]
[by Codex] Thanks, the conversion is not unconditional across SGLang backends. It is scoped to `FlashinferDispatcher`, which is instantiated only for `--moe-a2a-backend flashinfer`; DeepEP, NIXL, Mooncake, and standard dispatchers are unchanged.

The FlashInfer 0.6.12 `MoeAlltoAll` contract documents `token_selected_experts` as int32, and this dispatcher already sizes both the workspace and ID payload as `router_topk * 4`. Passing the router-produced int64 tensor therefore violated that boundary. The three MoE runners SGLang permits with FlashInfer A2A are also int32-compatible: FlashInfer Cutlass casts IDs with `.to(torch.int)`, CuteDSL casts to int32, and the TRT-LLM runner `PackTopkIds` explicitly asserts int32.

The patch normalizes IDs only at that A2A boundary and returns the received IDs as int32 for those runners. The focused standard and bypassed-top-k tests passed in the pinned nightly container (OCI job 168393), and the GLM-5 DEP2 E2E path subsequently passed. I added the ABI rationale beside the cast in b2642c7db8.

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py [hunk 1]
```diff
@@ -175,7 +179,11 @@ def dispatch(
         output_dtype = hidden_states.dtype
         x = hidden_states
         x_sf = None
-        topk_ids = topk_output.topk_ids
+        if TopKOutputChecker.format_is_bypassed(topk_output):
+            topk_output = topk_output.to_standard()
+        # FlashInfer MoeAlltoAll's expert-ID ABI is int32. This dispatcher is
```
Comment (nvpohanh) [inline python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:186]: Without modifying the code or comments, update the PR description with links to the FlashInfer MoeAlltoAll's expert-ID API and the docstring so that we can easily verify that FlashInfer indeed expects int32 for selected expert ids

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py [hunk 2]
```diff
@@ -175,7 +179,11 @@ def dispatch(
         output_dtype = hidden_states.dtype
         x = hidden_states
         x_sf = None
-        topk_ids = topk_output.topk_ids
+        if TopKOutputChecker.format_is_bypassed(topk_output):
```
Comment (nvpohanh) [inline python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:184]: Explain to me without changing the code: what does this line mean? What is `format_is_bypassed()` and what does `to_standard()` do?

File: test/registered/unit/layers/moe/test_flashinfer_dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    StandardTopKOutput,
+    TopKConfig,
+)
+from sglang.test.ci.ci_register import register_cpu_ci
+from sglang.test.test_utils import CustomTestCase
+
+register_cpu_ci(est_time=5, suite="base-a-test-cpu")
+
+
+class TestFlashinferDispatcher(CustomTestCase):
```
Comment (nvpohanh) [inline test/registered/unit/layers/moe/test_flashinfer_dispatcher.py]: I don't like this test. It doesn't actually run anything on GPU.

Could you check the existing GLM-5.2 accuracy tests? Tell me where the GLM-5.2 accuracy tests are located and whether we can add one more test case there with moe-a2a-backend set to `flashinfer` to cover the current coverage hole?

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py [hunk 4]
```diff
@@ -175,7 +175,7 @@ def dispatch(
         output_dtype = hidden_states.dtype
         x = hidden_states
         x_sf = None
-        topk_ids = topk_output.topk_ids
+        topk_ids = topk_output.topk_ids.to(torch.int32)
```
[the remark this answers is not in the document]
Comment (nvpohanh) [reply python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:188]: [by Codex] Addressed in ffaa7c6a4f. The dispatcher now recognizes `BypassedTopKOutput`, materializes it with `to_standard()`, and converts the resulting selected expert IDs to `int32` before passing them to FlashInfer.

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py (hunk 1 above)
Comment (nvpohanh) [reply python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:186]: [by Codex] Added the requested pinned FlashInfer 0.6.12 links to the PR description: `MoeAlltoAll.dispatch` and the downstream `moe_a2a_dispatch` docstring that specifies `token_selected_experts` is `int32`.

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py (hunk 2 above)
Comment (nvpohanh) [reply python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:184]: [by Codex] `format_is_bypassed()` is a TypeGuard implemented as `isinstance(topk_output, BypassedTopKOutput)`. A bypassed output defers routing and carries hidden states, router logits, and routing configuration instead of explicit `topk_ids`/`topk_weights`. `to_standard()` calls `select_experts(...)` to materialize those routing tensors as `StandardTopKOutput`, which FlashInfer A2A needs before dispatch. No code or comment change was made for this explanation.

File: test/registered/unit/layers/moe/test_flashinfer_dispatcher.py (hunk 3 above)
Comment (nvpohanh) [reply test/registered/unit/layers/moe/test_flashinfer_dispatcher.py]: [by Codex] The existing GLM-5.2 accuracy coverage is `test/registered/8-gpu-models/test_glm52_fp8.py`, in the nightly 8-GPU H200/B200 suite. I replaced the mocked CPU test with a TP8+DP8+FlashInferA2A variant there. `--enable-dp-attention` plus DP=TP satisfies FlashInfer A2A requirements; server validation then sets EP=TP, so the real `FlashinferDispatcher` runs on GPU.

**Comment by nvpohanh** [comment]
convert to draft for now since I need to verify that the added GLM-5.2 test could actually catch this failure

File: python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py (hunk 2 above)
Comment (nvpohanh) [inline python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:184]: Add a short comment about why this is needed

Comment (nvpohanh) [reply python/sglang/srt/layers/moe/token_dispatcher/flashinfer.py:184]: [by Codex] Addressed in 6f4239e39e. Added a short comment explaining that bypassed top-k output must be materialized because FlashInfer dispatch requires explicit top-k IDs and weights.

**Comment by leejnau** [comment]
> @leejnau @nvjullin @trevor-m Agents told me that this patch is needed to run GLM-5 + FlashInfer A2A. Do you think it makes sense?

my agents agree 🕵️

**Comment by nvpohanh** [comment]
@b8zhong Could you help me to review this? This fixes a topk_id dtype mismatch issue when using flashinfer alltoall. Thanks!

All the NV pipelines have passed.

**Review by b8zhong — APPROVED** [review]
/rerun-test test/registered/4-gpu-models/test_glm52_nvfp4.py

**Comment by nvpohanh** [comment]
@b8zhong is this PR okay to merge? Or do we need more reviews? Thanks!

File: test/registered/4-gpu-models/test_glm52_nvfp4.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import unittest
+from unittest.mock import patch
+
+from sglang.test.accuracy_test_runner import AccuracyTestParams
+from sglang.test.ci.ci_register import register_cuda_ci
+from sglang.test.performance_test_runner import PerformanceTestParams
+from sglang.test.run_combined_tests import run_combined_tests
+from sglang.test.test_utils import CustomTestCase, ModelLaunchSettings
+
+register_cuda_ci(est_time=1800, suite="nightly-4-gpu-b200", nightly=True)
```
Comment (Fridge003) [inline test/registered/4-gpu-models/test_glm52_nvfp4.py]: Can we make it a subtest of test/registered/ep/test_flashinfer_a2a.py instead
A nightly test with 1800s is too heavy

Comment (nvpohanh) [reply test/registered/4-gpu-models/test_glm52_nvfp4.py]: makes sense. Will update it

Comment (nvpohanh) [reply test/registered/4-gpu-models/test_glm52_nvfp4.py]: [by Codex] Addressed in 51678e2a5a. I moved the GLM-5.2 NVFP4 regression coverage into test/registered/ep/test_flashinfer_a2a.py as TestFlashinferA2ACutedslStaticFP4 and removed the standalone 1,800-second nightly accuracy/performance matrix. The consolidated case exercises Static EP, flashinfer_cutedsl, and FlashInfer A2A, then performs one generation request. Validated on AIHub B200x4 (job 1379463): TestFlashinferA2ACutedslStaticFP4.test_generate passed — 1 test in 471.443s, OK. The branch is also rebased onto latest origin/main at c00131ebaa.

**Comment by nvpohanh** [comment]
@b8zhong @Fridge003 I have updated the test definition. Could you review again? thanks!

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
Successful run: https://github.com/sgl-project/sglang/actions/runs/29388844304/job/87299509093?pr=29929
